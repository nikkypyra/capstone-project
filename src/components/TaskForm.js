import React, { useState } from 'react'
import styled from 'styled-components/macro'
import Header from './Header'
import SubmitButton from './SubmitButton'

export default function TaskForm() {
  const [description, setDescription] = useState('')
  const [date, setDate] = useState('')
  const [time, setTime] = useState('')
  const [person, setPerson] = useState('')

  const handleSubmit = (event) => {
    console.log(`
      Description: ${description}
      Date: ${date}
      Time: ${time}
      Person: ${person}
    `)

    event.preventDefault()
  }

  return (
    <>
      <Header />
      <BorderStyled>
        <Form onSubmit={handleSubmit}>
          <div className="description">
            <label>
              Description*
              <input
                type="text"
                value={description}
                maxLength="100"
                placeholder="Insert description"
                onChange={(e) => setDescription(e.target.value)}
                required
              />
            </label>
          </div>
          <div className="date">
            <label>
              Date*
              <input
                type="date"
                value={date}
                onChange={(e) => setDate(e.target.value)}
                required
              />
            </label>
          </div>
          <div className="time">
            <label>
              Time*
              <input
                type="time"
                value={time}
                onChange={(e) => setTime(e.target.value)}
                required
              />
            </label>
          </div>
          <div className="person">
            <label>
              To be completed by*
              <input
                type="text"
                value={person}
                maxLength="100"
                placeholder="Insert person to complete task"
                onChange={(e) => setPerson(e.target.value)}
                required
              />
            </label>
          </div>
          <SubmitButton />
        </Form>
        <p>*Mandatory Fields</p>
      </BorderStyled>
    </>
  )
}

const Form = styled.form`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr 1fr 1fr;
  align-items: center;
  grid-gap: 40px;
  color: var(--secondary);

  label,
  input {
    margin: 8px 0;
  }

  label {
    font-size: 18px;
  }
  input {
    width: 100%;
    height: 2rem;
    font-size: 16px;
    font-family: sans-serif;
    border: none;
    border-bottom: 1px solid var(--primary);
  }

  input[type='date']::selection {
    background-color: var(--primary);
  }

  .description {
    grid-row: 1/2;
    grid-column: span 2;
  }

  .date {
    grid-row: 2/3;
    grid-column: 1/2;
  }

  .date::selection {
    background: var(--primary);
  }

  .time {
    grid-row: 2/3;
    grid-column: 2/3;
  }

  .person {
    grid-row: 3/4;
    grid-column: span 2;
  }

  button {
    grid-row: 4/5;
    grid-column: span 2;
  }
`

const BorderStyled = styled.main`
  margin: 20px;
  padding: 12px;
  border: 4px solid var(--tertiary);
  border-radius: 12px;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.16);

  p {
    margin-top: 4px;
  }
`
