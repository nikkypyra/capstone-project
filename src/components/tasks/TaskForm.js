import React from 'react'
import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import SubmitButton from '../buttons/SubmitButton'
import CancelButton from '../buttons/CancelButton'

TaskForm.propTypes = {
  description: PropTypes.string.isRequired,
  setDescription: PropTypes.func.isRequired,
  date: PropTypes.string.isRequired,
  setDate: PropTypes.func.isRequired,
  time: PropTypes.string.isRequired,
  setTime: PropTypes.func.isRequired,
  person: PropTypes.string.isRequired,
  setPerson: PropTypes.func.isRequired,
  disabled: PropTypes.bool.isRequired,
  pet: PropTypes.object.isRequired,
}

export default function TaskForm({
  description,
  setDescription,
  date,
  setDate,
  time,
  setTime,
  person,
  setPerson,
  disabled,
  pet,
}) {
  return (
    <>
      <Cancel>
        <Link to={`/pet/${pet.id}`}>
          <CancelButton />
        </Link>
      </Cancel>
      <Description>
        <label>
          Description*
          <input
            type="text"
            name="description"
            value={description}
            maxLength="100"
            placeholder="Insert description"
            onChange={(e) => setDescription(e.target.value)}
            required
            autoFocus
          />
        </label>
      </Description>
      <Date>
        <label>
          Date*
          <input
            type="date"
            name="date"
            value={date}
            onChange={(e) => setDate(e.target.value)}
            required
          />
        </label>
      </Date>
      <Time>
        <label>
          Time*
          <input
            type="time"
            name="time"
            value={time}
            onChange={(e) => setTime(e.target.value)}
            required
          />
        </label>
      </Time>
      <Person>
        <label>
          To be completed by*
          <input
            type="text"
            name="person"
            value={person}
            maxLength="100"
            placeholder="Insert person to complete task"
            onChange={(e) => setPerson(e.target.value)}
            required
          />
        </label>
      </Person>
      <ButtonWrapper>
        <SubmitButton text="Submit" disabled={disabled} type="submit" />
      </ButtonWrapper>
    </>
  )
}

const Cancel = styled.div`
  grid-row: 1/2;
  grid-column: 2/3;
  display: flex;
  justify-content: flex-end;
`

const Description = styled.div`
  grid-row: 2/3;
  grid-column: span 2;
`

const Date = styled.div`
  grid-row: 3/4;
  grid-column: 1/2;
`

const Time = styled.div`
  grid-row: 3/4;
  grid-column: 2/3;
`

const Person = styled.div`
  grid-row: 4/5;
  grid-column: span 2;
`

const ButtonWrapper = styled.div`
  grid-row: 5/6;
  grid-column: span 2;
  button {
    width: 100%;
  }
`
