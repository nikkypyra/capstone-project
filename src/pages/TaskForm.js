import React, { useState } from 'react'
import styled from 'styled-components/macro'
import SubmitButton from '../components/Buttons/SubmitButton'
import CancelButton from '../components/Buttons/CancelButton'
import { v4 as uuidv4 } from 'uuid'
import { useHistory, useParams, Link } from 'react-router-dom'
import PropTypes from 'prop-types'

TaskForm.propTypes = {
  pets: PropTypes.array.isRequired,
  addTask: PropTypes.func.isRequired,
}

export default function TaskForm({ pets, addTask }) {
  const [description, setDescription] = useState('')
  const [date, setDate] = useState('')
  const [time, setTime] = useState('')
  const [person, setPerson] = useState('')
  const history = useHistory()
  const uniqueTaskId = uuidv4()
  const params = useParams()
  const pet = pets.find((pet) => pet.id === params.id)

  function handleSubmit(event) {
    event.preventDefault()
    addTask({
      description,
      date,
      time,
      person,
      complete: false,
      id: uniqueTaskId,
      petId: pet.id,
      petName: pet.name,
    })
    history.push(`/pet/${pet.id}`)
  }

  return (
    <>
      <main>
        <Form onSubmit={handleSubmit}>
          <div className="cancel">
            <Link to={`/pet/${pet.id}`}>
              <CancelButton />
            </Link>
          </div>
          <div className="description">
            <label>
              Description*
              <input
                type="text"
                value={description}
                maxLength="100"
                placeholder="Insert description"
                onChange={(e) => setDescription(e.target.value)}
                required
                autoFocus
              />
            </label>
          </div>
          <div className="date">
            <label>
              Date*
              <input
                type="date"
                value={date}
                onChange={(e) => setDate(e.target.value)}
                required
              />
            </label>
          </div>
          <div className="time">
            <label>
              Time*
              <input
                type="time"
                value={time}
                onChange={(e) => setTime(e.target.value)}
                required
              />
            </label>
          </div>
          <div className="person">
            <label>
              To be completed by*
              <input
                type="text"
                value={person}
                maxLength="100"
                placeholder="Insert person to complete task"
                onChange={(e) => setPerson(e.target.value)}
                required
              />
            </label>
          </div>
          <SubmitButton text="Submit" />
          <p>*Mandatory Fields</p>
        </Form>
      </main>
    </>
  )
}

const Form = styled.form`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 3fr 3fr 3fr 3fr 1fr;
  align-items: center;
  color: var(--secondary);
  margin: 20px;
  padding: 12px;
  border: 4px solid var(--tertiary);
  border-radius: 12px;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.16);

  label,
  input {
    margin: 8px 0;
  }

  label {
    font-size: 18px;
  }
  input {
    width: 100%;
    height: 2rem;
    font-size: 16px;
    font-family: sans-serif;
    border: none;
    border-bottom: 1px solid var(--primary);
  }

  input[type='date']::selection {
    background-color: var(--primary);
  }

  .cancel {
    grid-row: 1/2;
    grid-column: 2/3;
    display: flex;
    justify-content: flex-end;
  }

  .description {
    grid-row: 2/3;
    grid-column: span 2;
  }

  .date {
    grid-row: 3/4;
    grid-column: 1/2;
  }

  .time {
    grid-row: 3/4;
    grid-column: 2/3;
  }

  .person {
    grid-row: 4/5;
    grid-column: span 2;
  }

  button {
    grid-row: 5/6;
    grid-column: span 2;
  }

  p {
    grid-row: 6/7;
    margin-top: 4px;
  }
`
