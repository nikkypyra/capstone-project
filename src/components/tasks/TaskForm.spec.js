import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import TaskForm from './TaskForm'

afterEach(cleanup)
const description = {}
const setDescription = jest.fn()
const date = {}
const setDate = jest.fn()
const time = {}
const setTime = jest.fn()
const person = {}
const setPerson = jest.fn()
const disabled = true
const pet = {}

describe('<TaskForm />', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <TaskForm
          description={description}
          setDescription={setDescription}
          date={date}
          setDate={setDate}
          time={time}
          setTime={setTime}
          person={person}
          setPerson={setPerson}
          disabled={disabled}
          pet={pet}
        />
      </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders placeholder text in first input field', () => {
    const { getByPlaceholderText } = render(
      <MemoryRouter>
        <TaskForm
          description={description}
          setDescription={setDescription}
          date={date}
          setDate={setDate}
          time={time}
          setTime={setTime}
          person={person}
          setPerson={setPerson}
          disabled={disabled}
          pet={pet}
        />
      </MemoryRouter>
    )
    expect(getByPlaceholderText(/insert description/i)).toBeInTheDocument()
  })

  test('renders placeholder text in last input field', () => {
    const { getByPlaceholderText } = render(
      <MemoryRouter>
        <TaskForm
          description={description}
          setDescription={setDescription}
          date={date}
          setDate={setDate}
          time={time}
          setTime={setTime}
          person={person}
          setPerson={setPerson}
          disabled={disabled}
          pet={pet}
        />
      </MemoryRouter>
    )
    expect(
      getByPlaceholderText(/insert person to complete task/i)
    ).toBeInTheDocument()
  })
})
