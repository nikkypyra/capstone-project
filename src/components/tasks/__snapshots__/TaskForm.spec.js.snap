// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskForm /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="TaskForm__Cancel-sc-2ywoyi-0 gdkiwY"
  >
    <a
      href="/pet/undefined"
    >
      <div
        class="sc-AxjAm juTgyb"
      >
        <img
          alt="cancel"
          src="cancel.png"
        />
      </div>
    </a>
  </div>
  <div
    class="TaskForm__Description-sc-2ywoyi-1 cSeoRp"
  >
    <label>
      Description*
      <input
        maxlength="100"
        name="description"
        placeholder="Insert description"
        required=""
        type="text"
        value="[object Object]"
      />
    </label>
  </div>
  <div
    class="TaskForm__Date-sc-2ywoyi-2 bUpfGE"
  >
    <label>
      Date*
      <input
        name="date"
        required=""
        type="date"
        value="[object Object]"
      />
    </label>
  </div>
  <div
    class="TaskForm__Time-sc-2ywoyi-3 jIICJW"
  >
    <label>
      Time*
      <input
        name="time"
        required=""
        type="time"
        value="[object Object]"
      />
    </label>
  </div>
  <div
    class="TaskForm__Person-sc-2ywoyi-4 iEaioL"
  >
    <label>
      To be completed by*
      <input
        maxlength="100"
        name="person"
        placeholder="Insert person to complete task"
        required=""
        type="text"
        value="[object Object]"
      />
    </label>
  </div>
  <div
    class="TaskForm__ButtonWrapper-sc-2ywoyi-5 dHbYv"
  >
    <button
      class="SubmitButton__Button-sc-1s0oo90-0 beZJxO"
      disabled=""
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
